src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts

```diff
@@ -0,0 +1,777 @@
+    width: number;
+    height: number;
+    sampleCount: number;
+    nearFar: readonly number[];
+    clipSpacePoints: readonly number[][];
+    interStagePoints: readonly number[][];
+    outputCode: string;
+  }
+) {
+  const interpolate = `${interpolationType}${
+    interpolationSampling ? `, ${interpolationSampling}` : ''
+  }`;
+  const module = t.device.createShaderModule({
+    code: `
+      struct Uniforms {
+        resolution: vec2f,
+      };
+
+      @group(0) @binding(0) var<uniform> uni: Uniforms;
+
+      struct Vertex {
+        @builtin(position) position: vec4f,
+        @location(0) @interpolate(${interpolate}) interpolatedValue: vec4f,
+      };
+
+      @vertex fn vs(@builtin(vertex_index) vNdx: u32) -> Vertex {
+        let pos = array(
+          ${clipSpacePoints.map(p => `vec4f(${p.join(', ')})`).join(', ')}
+        );
+        let interStage = array(
+          ${interStagePoints.map(p => `vec4f(${p.join(', ')})`).join(', ')}
+        );
+        var v: Vertex;
+        v.position = pos[vNdx];
+        v.interpolatedValue = interStage[vNdx];
+        _ = uni;
+        return v;
+      }
+
+      struct FragOut {
+        @location(0) out0: vec4f,
+        @location(1) out1: vec4f,
+        @location(2) out2: vec4f,
+        @location(3) out3: vec4f,
+      };
+
+      fn u32ToRGBAUnorm(u: u32) -> vec4f {
+        return vec4f(
+          f32((u >> 24) & 0xFF) / 255.0,
+          f32((u >> 16) & 0xFF) / 255.0,
+          f32((u >>  8) & 0xFF) / 255.0,
+          f32((u >>  0) & 0xFF) / 255.0,
+        );
+      }
+
+      @fragment fn fs(vin: Vertex) -> FragOut {
+        var f: FragOut;
+        let v = ${outputCode};
+        let u = bitcast<vec4u>(v);
+        f.out0 = u32ToRGBAUnorm(u[0]);
+        f.out1 = u32ToRGBAUnorm(u[1]);
+        f.out2 = u32ToRGBAUnorm(u[2]);
+        f.out3 = u32ToRGBAUnorm(u[3]);
+        _ = vin.interpolatedValue;
+        return f;
+      }
+    `,
+  });
+
+  const textures = range(4, () => {
+    const texture = t.device.createTexture({
+      size: [width, height],
+      usage:
+        GPUTextureUsage.RENDER_ATTACHMENT |
+        GPUTextureUsage.TEXTURE_BINDING |
+        GPUTextureUsage.COPY_SRC,
+      format: 'rgba8unorm',
+      sampleCount,
+    });
+    t.trackForCleanup(texture);
+    return texture;
+  });
+
+  const pipeline = t.device.createRenderPipeline({
+    layout: 'auto',
+    vertex: {
+      module,
+      entryPoint: 'vs',
+    },
+    fragment: {
+      module,
+      entryPoint: 'fs',
+      targets: textures.map(() => ({ format: 'rgba8unorm' })),
+    },
+    multisample: {
+      count: sampleCount,
+    },
+  });
+
+  const texture0 = textures[0];
+  const uniformBuffer = t.device.createBuffer({
+    size: 8,
+    usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
+  });
+  t.trackForCleanup(uniformBuffer);
+  t.device.queue.writeBuffer(uniformBuffer, 0, new Float32Array([texture0.width, texture0.height]));
```

Comment:
nit. Maybe could get rid of `texture0` and just use `width` and `height` instead of `texture0.width` aand `texture0.height`?